plugins/inventory/gql_inventory.py

```diff
@@ -302,7 +302,10 @@ def parse(self, inventory, loader, path, cache=True):
         }
         if token:
             self.headers.update({"Authorization": "Token %s" % token})
-
+        else:
```

Comment:
We can remove this code and change L44 to `required: True`

Reply:
If we put `required: True` then the key will need to be there. It would not allow us to use the environment var loading of the token.

Reply:
I ran it locally, but I am on Ansible 2.10:

`inventory.yml`

```
---
plugin: networktocode.nautobot.gql_inventory
api_endpoint: http://localhost:8000
validate_certs: True
query:
  interfaces: name
additional_variables:
  - interfaces
```

```
❯ unset NAUTOBOT_TOKEN
❯ ansible-inventory -vvvv --list --inventory inventory.yml
ansible-inventory 2.10.15
  config file = /Users/myohman/cloned-repos/nautobot-ansible/ansible.cfg
  configured module search path = ['/Users/myohman/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible
  executable location = /Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/bin/ansible-inventory
  python version = 3.8.6 (default, Dec 27 2020, 10:06:23) [Clang 12.0.0 (clang-1200.0.32.28)]
Using /Users/myohman/cloned-repos/nautobot-ansible/ansible.cfg as config file
setting up inventory plugins
Loading collection networktocode.nautobot from /Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot
[WARNING]:  * Failed to parse /Users/myohman/cloned-repos/nautobot-ansible/inventory.yml with ansible_collections.networktocode.nautobot.plugins.inventory.inventory plugin: Incorrect plugin name in file:
networktocode.nautobot.gql_inventory
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/inventory/manager.py", line 289, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot/plugins/inventory/inventory.py", line 1350, in parse
    self._read_config_data(path=path)
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/plugins/inventory/__init__.py", line 227, in _read_config_data
    raise AnsibleParserError("Incorrect plugin name in file: %s" % config.get('plugin', 'none found'))
[WARNING]:  * Failed to parse /Users/myohman/cloned-repos/nautobot-ansible/inventory.yml with ansible_collections.networktocode.nautobot.plugins.inventory.gql_inventory plugin: No setting was provided for
required configuration plugin_type: inventory plugin: ansible_collections.networktocode.nautobot.plugins.inventory.gql_inventory setting: token
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/inventory/manager.py", line 289, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot/plugins/inventory/gql_inventory.py", line 293, in parse
    self._read_config_data(path=path)
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/plugins/inventory/__init__.py", line 232, in _read_config_data
    self.set_options(direct=config)
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/plugins/__init__.py", line 75, in set_options
    self._options = C.config.get_plugin_options(get_plugin_class(self), self._load_name, keys=task_keys, variables=var_options, direct=direct)
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/config/manager.py", line 361, in get_plugin_options
    options[option] = self.get_config_value(option, plugin_type=plugin_type, plugin_name=name, keys=keys, variables=variables, direct=direct)
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/config/manager.py", line 429, in get_config_value
    value, _drop = self.get_config_value_and_origin(config, cfile=cfile, plugin_type=plugin_type, plugin_name=plugin_name,
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/config/manager.py", line 517, in get_config_value_and_origin
    raise AnsibleError("No setting was provided for required configuration %s" %
[WARNING]: Unable to parse /Users/myohman/cloned-repos/nautobot-ansible/inventory.yml as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
{
    "_meta": {
        "hostvars": {}
    },
    "all": {
        "children": [
            "ungrouped"
        ]
    }
}
```
With exporting the token.
```
❯ export NAUTOBOT_TOKEN=0123456789abcdef0123456789abcdef01234567
❯ ansible-inventory -vvvv --list --inventory inventory.yml
ansible-inventory 2.10.15
  config file = /Users/myohman/cloned-repos/nautobot-ansible/ansible.cfg
  configured module search path = ['/Users/myohman/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible
  executable location = /Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/bin/ansible-inventory
  python version = 3.8.6 (default, Dec 27 2020, 10:06:23) [Clang 12.0.0 (clang-1200.0.32.28)]
Using /Users/myohman/cloned-repos/nautobot-ansible/ansible.cfg as config file
setting up inventory plugins
Loading collection networktocode.nautobot from /Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot
Replacing invalid character(s) "{' '}" in group name (Test Site2)
[WARNING]: Invalid characters were found in group names and automatically replaced, use -vvvv to see details
[WARNING]:  * Failed to parse /Users/myohman/cloned-repos/nautobot-ansible/inventory.yml with ansible_collections.networktocode.nautobot.plugins.inventory.inventory plugin: Incorrect plugin name in file:
networktocode.nautobot.gql_inventory
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/inventory/manager.py", line 289, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot/plugins/inventory/inventory.py", line 1350, in parse
    self._read_config_data(path=path)
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/plugins/inventory/__init__.py", line 227, in _read_config_data
    raise AnsibleParserError("Incorrect plugin name in file: %s" % config.get('plugin', 'none found'))
[WARNING]:  * Failed to parse /Users/myohman/cloned-repos/nautobot-ansible/inventory.yml with ansible_collections.networktocode.nautobot.plugins.inventory.gql_inventory plugin: Could not find group Test
Site2 in inventory
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/inventory/manager.py", line 289, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot/plugins/inventory/gql_inventory.py", line 315, in parse
    self.main()
  File "/Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot/plugins/inventory/gql_inventory.py", line 271, in main
    self.create_inventory(key, device["name"])
  File "/Users/myohman/.ansible/collections/ansible_collections/networktocode/nautobot/plugins/inventory/gql_inventory.py", line 182, in create_inventory
    self.inventory.add_host(host, group)
  File "/Users/myohman/Library/Caches/pypoetry/virtualenvs/nautobot-ansible-modules-DR2vR1FY-py3.8/lib/python3.8/site-packages/ansible/inventory/data.py", line 201, in add_host
    raise AnsibleError("Could not find group %s in inventory" % group)
[WARNING]: Unable to parse /Users/myohman/cloned-repos/nautobot-ansible/inventory.yml as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
{
    "_meta": {
        "hostvars": {}
    },
    "all": {
        "children": [
            "ungrouped"
        ]
    }
}
```
It errors out on something else, but fails properly without setting the token within the inventory file and works fine if it's only defined via the environment variable.

Reply:
I tested and experienced the same behavior. Code updated to reflect that.